File: src/server/src/routes/discord/server/role-categories.ts
import { Hono } from "hono";
import { getCookie } from "hono/cookie";
import { HTTPException } from "hono/http-exception";
import {
	createRoleCategory,
	deleteRoleCategory,
	updateRoleCategoryOfRole,
} from "../../../functions/db.js";
import { authenticateAndAuthorizeUser } from "../../../functions/discord-bot.js";

const app = new Hono();

app.post("/create", async (c) => {
	const serverId = c.req.param("serverId");
	const body = await c.req.json();
	const accessToken = getCookie(c, "discord_access_token");
	if (!body.name) {
		throw new HTTPException(400, {
			message: "field 'name' is required in body.",
		});
	}
	const { manageGuildPermission } = await authenticateAndAuthorizeUser(
		serverId as string,
		accessToken as string,
		true,
	);
	if (manageGuildPermission) {
		const roleCategory = await createRoleCategory({
			serverId: serverId as string,
			name: body.name as string,
		});
		return c.json(roleCategory, 200);
	}
	throw new HTTPException(403, {
		message: "User does not have manage server permission.",
	});
});

app.delete("/:roleCategoryId", async (c) => {
	const serverId = c.req.param("serverId");
	const roleCategoryId = c.req.param("roleCategoryId");
	const accessToken = getCookie(c, "discord_access_token");
	const { manageGuildPermission } = await authenticateAndAuthorizeUser(
		serverId as string,
		accessToken as string,
		true,
	);
	if (manageGuildPermission) {
		const roleCategory = await deleteRoleCategory({
			roleCategoryId: roleCategoryId as unknown as number,
			serverId: serverId as string,
		});
		return c.json(roleCategory, 200);
	}
	throw new HTTPException(403, {
		message: "User does not have manage server permission.",
	});
});

app.post("/:roleCategoryId/assign", async (c) => {
	const serverId = c.req.param("serverId");
	const roleCategoryId = c.req.param("roleCategoryId");
	const body = await c.req.json();
	const accessToken = getCookie(c, "discord_access_token");
	const { manageGuildPermission } = await authenticateAndAuthorizeUser(
		serverId as string,
		accessToken as string,
		true,
	);
	if (manageGuildPermission) {
		const roleCategory = await updateRoleCategoryOfRole({
			roleId: body.roleId as string,
			roleCategoryId: roleCategoryId as unknown as number,
			serverId: serverId as string,
		});
		return c.json(roleCategory, 200);
	}
	throw new HTTPException(403, {
		message: "User does not have manage server permission.",
	});
});

export default app;

File: src/server-rust/src/routes/role_categories.rs
//! Port of `../server/src/routes/discord/server/role-categories.ts`.
//!
//! Nested under `/discord/servers/{serverId}/role-categories`.

use axum::extract::{Path, State};
use axum::http::StatusCode;
use axum::response::IntoResponse;
use axum::routing::{delete, post};
use axum::{Json, Router};
use axum_extra::extract::CookieJar;

use crate::db;
use crate::discord::bot;
use crate::error::{HttpError, Result};
use crate::models::{AssignRoleCategoryToRoleRequestBody, NameRequiredRequestBody};
use crate::state::AppState;
use crate::validation::{is_valid_discord_id, validated_name, CATEGORY_NAME_MAX};

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/create", post(create_role_category))
        .route("/{roleCategoryId}", delete(delete_role_category))
        .route("/assign", post(assign_role_category))
}

fn access_token(jar: &CookieJar) -> String {
    jar.get("discord_access_token")
        .map(|c| c.value().to_string())
        .unwrap_or_default()
}

async fn create_role_category(
    State(state): State<AppState>,
    Path(server_id): Path<String>,
    jar: CookieJar,
    Json(body): Json<NameRequiredRequestBody>,
) -> Result<impl IntoResponse> {
    let token = access_token(&jar);
    let authed =
        bot::authenticate_and_authorize_user(&state, &server_id, &token, true, true).await?;
    if authed.manage_guild_permission {
        let name = validated_name(&body.name, CATEGORY_NAME_MAX, "Name")?;
        let role_category = db::create_role_category(&state.db, &server_id, &name, false).await?;
        return Ok((StatusCode::OK, Json(role_category)));
    }
    Err(HttpError::forbidden(
        "User does not have manage server permission.",
    ))
}

async fn delete_role_category(
    State(state): State<AppState>,
    Path((server_id, role_category_id)): Path<(String, i32)>,
    jar: CookieJar,
) -> Result<impl IntoResponse> {
    let token = access_token(&jar);
    let authed =
        bot::authenticate_and_authorize_user(&state, &server_id, &token, true, true).await?;
    if authed.manage_guild_permission {
        let role_category =
            db::delete_role_category(&state.db, role_category_id, &server_id).await?;
        return Ok((StatusCode::OK, Json(role_category)));
    }
    Err(HttpError::forbidden(
        "User does not have manage server permission.",
    ))
}

async fn assign_role_category(
    State(state): State<AppState>,
    Path(server_id): Path<String>,
    jar: CookieJar,
    Json(body): Json<AssignRoleCategoryToRoleRequestBody>,
) -> Result<impl IntoResponse> {
    let token = access_token(&jar);
    let authed =
        bot::authenticate_and_authorize_user(&state, &server_id, &token, true, true).await?;
    if authed.manage_guild_permission {
        if !is_valid_discord_id(&body.role_id) {
            return Err(HttpError::bad_request("Invalid role ID."));
        }
        if body.role_category_id.is_some_and(|id| id <= 0) {
            return Err(HttpError::bad_request("Invalid role category ID."));
        }
        let role_category = db::update_role_category_of_role(
            &state.db,
            &body.role_id,
            body.role_category_id,
            &server_id,
        )
        .await?;
        return Ok((StatusCode::OK, Json(role_category)));
    }
    Err(HttpError::forbidden(
        "User does not have manage server permission.",
    ))
}
